examples/reduce.c

```diff
@@ -0,0 +1,412 @@
+	char msg[PTL_EV_STR_SIZE];
+	ptl_event_t ev;
+
+	/* sending the content of mdX to meY */
+	int ret = PtlMDBind(nih, &(ranks[rank].md), &mdh);
+	if (ret != PTL_OK) {
+		fprintf(stderr, "PtlMDBind failed : %s \n", PtlToStr(ret, PTL_STR_ERROR));
+		return 1;
+	}
+
+	ret = PtlAtomic(mdh, 0, ranks[rank].size, PTL_ACK_REQ, id, ranks[rank].pti, 0, 0, NULL, 0,
+			PTL_SUM, PTL_UINT32_T);
+	if (ret != PTL_OK) {
+		fprintf(stderr, "PtlAtomic failed : %s \n", PtlToStr(ret, PTL_STR_ERROR));
+		return 1;
+	} else {
+		ret = PtlCTInc(ranks[rank].counter, increment);
+		if (ret != PTL_OK) {
+			fprintf(stderr, "PtlCTInc failed : %s \n", PtlToStr(ret, PTL_STR_ERROR));
+			return 1;
+		}
+	}
+
+	ret = PtlEQWait(eqh, &ev);
+	if (ret != PTL_OK) {
+		fprintf(stderr, "PtlEQWait failed : %s \n", PtlToStr(ret, PTL_STR_ERROR));
+		return 1;
+	}
+	PtlEvToStr(0, &ev, msg);
+	printf("Event : %s", msg);
+
+	ret = PtlEQWait(eqh, &ev);
+	if (ret != PTL_OK) {
+		fprintf(stderr, "PtlEQWait failed : %s \n", PtlToStr(ret, PTL_STR_ERROR));
+		return 1;
+	}
+	PtlEvToStr(0, &ev, msg);
+	printf("Event : %s", msg);
+
+	ret = PtlEQWait(eqh, &ev);
+	if (ret != PTL_OK) {
+		fprintf(stderr, "PtlEQWait failed : %s \n", PtlToStr(ret, PTL_STR_ERROR));
+		return 1;
+	}
+	PtlEvToStr(0, &ev, msg);
+	printf("Event : %s", msg);
+
+	if (rank % 2 == 1) {
+		/* we copy the start me to the start md */
+		ranks[rank].md.start = ranks[rank].me.start;
+
+		ret = PtlMDBind(nih, &ranks[rank].md, &mdh);
+		if (ret != PTL_OK) {
+			fprintf(stderr, "PtlMDBind failed : %s \n", PtlToStr(ret, PTL_STR_ERROR));
+			return 1;
+		}
+
+		ret = PtlTriggeredAtomic(mdh, 0, ranks[rank].size, PTL_ACK_REQ, id, pti2, 0, 0,
+					 NULL, 0, PTL_SUM, PTL_UINT32_T, ranks[rank].counter, 2);
+		if (ret != PTL_OK) {
+			fprintf(stderr, "PtlAtomic failed : %s \n", PtlToStr(ret, PTL_STR_ERROR));
+			return 1;
+		}
+
+		ret = PtlEQWait(eqh, &ev);
+		if (ret != PTL_OK) {
+			fprintf(stderr, "PtlEQWait failed : %s \n", PtlToStr(ret, PTL_STR_ERROR));
+			return 1;
+		}
+		PtlEvToStr(0, &ev, msg);
+		printf("Event : %s", msg);
+
+		ret = PtlEQWait(eqh, &ev);
+		if (ret != PTL_OK) {
+			fprintf(stderr, "PtlEQWait failed : %s \n", PtlToStr(ret, PTL_STR_ERROR));
+			return 1;
+		}
+		PtlEvToStr(0, &ev, msg);
+		printf("Event : %s", msg);
+
+		ret = PtlEQWait(eqh, &ev);
+		if (ret != PTL_OK) {
+			fprintf(stderr, "PtlEQWait failed : %s \n", PtlToStr(ret, PTL_STR_ERROR));
+			return 1;
+		}
+		PtlEvToStr(0, &ev, msg);
+		printf("Event : %s", msg);
+	}
+
+	return 0;
+}
+
+int main(void)
+{
+	int ret;
+	ptl_handle_ni_t nih;
+	ptl_handle_eq_t eqh;
+	ptl_index_t pti0;
+	ptl_index_t pti1;
+	ptl_index_t pti2;
+	ptl_handle_le_t leh;
+	ptl_process_t id;
+	ptl_event_t ev;
+	ptl_handle_ct_t counter0;
+	ptl_handle_ct_t counter1;
+
+	/* describe the memory used to receive messages */
+	ptl_me_t me0;
+	ptl_me_t me1;
+	ptl_me_t me2;
+
+	/* describe the memory used to send messages */
+	ptl_md_t md0;
+	ptl_md_t md1;
+	ptl_md_t md2;
+	ptl_md_t md3;

```

Comment:
Why do we declare those at the top level? Each "rank" should manage it's own resources in it's own space I think